### PR TITLE
Fix end of file and EOB op-code

### PR DESCRIPTION
**Pull request by dinosaure**

### CODE REVIEW

**Comment by dinosaure** [comment]
This bug is ... big.

### The bug
To contextualize. It's possible that the EOB which signals the end of the block is not the larger _op-code_ given by the huffman tree. The problem is simple, we consider that at any steps, we should have, at least, 15 bits into our `hold` - because this is the largest _op-code_  possible for any huffman tree.

However, if we reach the end of the stream, it's possible to have less than 15 bits. In that case, we probably reach the end of the stream - but we are not sure, only (`i_rem d < 0` can asserts that we reach the end of the stream). Two possibilities comes:
1) if `i_rem d = 0`, we should ask the user to give to us more that what we have
2) if `i_rem d < 0`, it's possible to have the end of the block

The first case is easy and it is handled by `slow_inflate`/`c_peek_bits` which re-asks 15 bits to the user. The second case is more tricky because: even if we really reach the end of the stream, that does not mean we have the `EOB` _op-code_. We still must check it. `c_peek_bits` does that in this specific case:
1) it tests if the current value of `hold` is the `EOB` _op-code_ (`== 256`)
2) it tests if we have enough bits to say that it's the `EOB` code

The second tests invalidate a specific situation which is well tested in our test suit. `hold` is a native integer and even if you have virtually 2 bits, we will get 15 bits to try to find an opcode. Remember the situation, we are in a context where we don't have enough to have 15 bits in our `hold` but we are at the end of the stream.

So `hold` can matches `EOB` even if you don't have enough bits to really matches `EOB`. Imagine that `EOB` is `0011`, `hold` with 2 bits (and left bits are set to `0`) can matches `EOB` but if you look on `bits` (which is equal to `2`), we don't have enough to really match `EOB` because we require 4 bits for instance.

The second test is mandatory to raise a proper error so. Then `d.last` means that we properly have the last block so we can safely execute the continuation which will terminates with `End_of_inflate`.

### The API
The definition of the API does not change. But the semantic or how we can deal with the stream API change a bit in some specific situations. As far as I can tell, this patch does not change anything with `ocaml-git`/`irmin` (which use the stream API). It unlocks some discovered bugs in some specific contexts (depending on the size of _chunk_, etc.).

At the end, the inflation is not so smart than before and need some helps to really finish the inflation of the stream. Some (higher) tests are updated in consequence when we must add a new state about the decoding (when we transmit directly to it the encoding) when we know that we will reach the end of the stream.

In other words, `Inf.src decoder t 0 0` becomes mandatory to really signals the end of the stream and help the decoder to choose the right direction between:
- the end of the chunk and a partial contents
- the end of the stream and the check of `EOB`

For an higher point of view, it should not change anything and `Zl` or `Gz` should be good - I did not notice errors on them. But for the user of `De` (on the stream API only), the user must rewrite a bit the loop to be sure to signal at the end the end of the stream - otherwise, it will have an undefined behavior.
